onnxruntime/core/providers/webnn/builders/helper.h

```diff
@@ -39,7 +41,35 @@ std::string GetShapeString(std::vector<T>& shape) {
   return shape_info.str();
 }
 
-bool GetShapeByTensor(const onnx::TensorProto& tensor, std::vector<int64_t>& shape, const logging::Logger& logger);
+template <typename T>
+bool ReadIntArrayFrom1DTensor(const onnx::TensorProto& tensor, std::vector<T>& array, const logging::Logger& logger) {
+  std::vector<uint8_t> unpacked_tensor;
+  auto status = onnxruntime::utils::UnpackInitializerData(tensor, unpacked_tensor);
+  if (!status.IsOK()) {
+    LOGS(logger, ERROR) << "Error while unpacking shape: " << status.ErrorMessage();
+    return false;
+  }
+  const auto& dims = tensor.dims();
+  if (dims.size() != 1) {
+    LOGS(logger, VERBOSE) << "The tensor must be 1D.";
+    return false;
+  }
+  int64_t rank = dims[0];
+  if (tensor.data_type() != ONNXTensorElementDataType::ONNX_TENSOR_ELEMENT_DATA_TYPE_INT64) {
+    LOGS(logger, VERBOSE) << "The type of tensor's element data must be INT64.";
+    return false;
+  }
+  const int64_t* array_data = reinterpret_cast<const int64_t*>(unpacked_tensor.data());
+  if (std::is_same<T, int64_t>::value) {
+    array.assign(array_data, array_data + rank);
+  } else if (std::is_same<T, int32_t>::value) {
+    std::vector<int64_t> raw_array = std::vector<int64_t>{array_data, array_data + rank};
+    std::transform(raw_array.cbegin(), raw_array.cend(),
+                   std::back_inserter(array),
+                   [](int64_t dim) -> int32_t { return SafeInt<int32_t>(dim); });
+  }
```

Comment:
```suggestion
  if constexpr (std::is_same<T, int64_t>::value) {
    array.assign(array_data, array_data + rank);
  } else if constexpr (std::is_same<T, int32_t>::value) {
    std::transform(array_data, array_data + rank,
                   std::back_inserter(array),
                   [](int64_t dim) -> int32_t { return SafeInt<int32_t>(dim); });
  }

```
- Since the template parameter is known at compile time, we should make the if statement constexpr.
- No need to allocate a temporary vector, since we're just returning the final one.

Reply:
Updated with your suggestions. Thanks!